### ISSUE TITLE
Ability to compose more than 22 arguments?

### ISSUE THREAD

**Issue by frgomes** [bkirwi/decline#38]
Hello,

I've done a simple PoC with ``decline`` and I'm stumbled on a difficulty or maybe on my own feet.
This is the scenario:

My application is made of multiple modules and the vast majority of them provide a command line parser.
I can pick and choose these modules so that I can build specialized ubber .jar files according to circumstances.

In a nutshell, I would like to have the ability of composing several command line parsers.

If I'm not mistaken, "the way composition works" (or at least: the way it is suggested in the documentation) is via tuples. However, the immediate concern that immediately comes to my mind is that tuples are limited to 22 elements (something which will only be lifted with Dotty).

Could someone suggest a way of composing parsers so that an arbitrary number of command line arguments would be possible?

Thanks

**Comment by frgomes** [comment]
Hello @bkirwi 
Thanks a lot for your prompt answer!

## Regarding `orElse`:

I've defined 50 args in a proof of concept and I've tried the `orElse` thing you mentioned.

```
val a1to10 = (a1,a2,a3,a4,a5,a6,a7,a8,a9,a10).mapN { 
(n1,n2,n3,n4,n5,n6,n7,n8,n9,n10) => println("1to10")
}
val a11to20 = (a11,a12,a13,a14,a15,a16,a17,a18,a19,a20).mapN { 
  (n1,n2,n3,n4,n5,n6,n7,n8,n9,n10) => println("11to20")
}
val a21to30 = (a21,a22,a23,a24,a25,a26,a27,a28,a29,a30).mapN { 
  (n21,n22,n23,n24,n25,n26,n27,n28,n29,n30) => println("21to30")
}
val t1 = a1to10 orElse a11to20 orElse a21to30
```

Yes, it works (as expected!).

## A tentative on a supposed `andThen` construct

It came to my mind that eventually the code had some sort of `andThen` available. Indeed, digging briefly inside the code I found that the parser uses some sort of ``andThen`` somehow, but I really spent any relevant effort trying to understand the details and nuances. I've then tried this:

```
val t2 = a1to10 andThen a11to20 andThen a21to30
  cmd7.sc:1: value andThen is not a member of com.monovore.decline.Opts[Unit]
  val t2 = a1to10 andThen a11to20 andThen a21to30
```

This code above does not compile (not really a surprise... since I know zero about the API!), but I would like to ask you if some sort of ``andThen`` is supported or exposed by the parser.

Using ``orElse`` makes sense when you are using a MySQL driver _or_ a Postgres Driver. However, when you are interested on a number of arguments _and_ some more arguments _and_ some other arguments... it would be interesting to have something which allows that.

Once again, thank you very much for your help! :-)

**Comment by bkirwi** [comment]
> when you are interested on a number of arguments plus some more arguments plus some other arguments...

For cases like that, the usual idiom is to use the `mapN` business as above. For example, this works:

```scala
val t2 = (a1to10, a11to20, a21to30).mapN { (n1to10, n11to20, n21to30) => println("1 to 30!) }
```

I'm definitely interested in improvements to the docs, or extensions to the API, that would make this stuff more accessible. (I'm not sure what the expectation you had for `andThen` was... if you have `optsA: Opts[A]` and `optsB: Opts[B]`, what type should `optsA andThen optsB` have? We should be able to find or add something that matches your intuitions.)

**Comment by DougC** [comment]
I was also looking for an `andThen` today. My scenario is that I have a `flag` and if it is set then I have other arguments to check. `andThen` is the applicative equivalent of `flatMap`, i.e. it lets you sequence things that are dependent on each other.

So, in simple terms, for `Opts[A]`, it would look like `def andThen[B](f: A => Opts[B])`

**Comment by bkirwi** [comment]
I don't think a `def andThen[B](f: A => Opts[B])` is possible for us to support. (Briefly: we don't have a `A` until _after_ parsing, but we need the `Opts[B]` _before_ parsing to know how to build the parser / render help / etc.)

However! The scenario you describe ought to be possible with the library as-is. One approach I like:

```scala
val flag = Opts.flag("flag", "...")
val option = Opts.option[Int]("option", "...")
val arg = Opts.argument[String]("positional")

val combined: Opts[Option[(Int, String)]] =
  (flag, option, arg).mapN { (_, int, string) => int -> string }.orNone
```

If the flag is set, the option and the arg are both required... and if the flag is unset, passing the option isn't allowed. You can usually capture dependencies like you describe in the "structure" of the parser, without needing to look at the values...

**Comment by DougC** [comment]
Yes, I was thinking about this overnight. `andThen` makes sense as a function at run time, but, as you point out, not for rendering the help as we can't know what the `Opts[B]` would be in a static context.

The `combined` example you provided is exactly the situation I have where I thought `andThen` would make sense!

**Comment by frgomes** [comment]
@bkirwi : I'd like to contribute with an example involving the ability to compose parsers.

```scala
import cats.implicits._
import com.monovore.decline._

case class Args(fileset: CmdFileset, encrypt: CmdEncrypt, opts: CmdOptions)
case class CmdFileset(chdir: String, label: Option[String], files: Seq[String])
case class CmdEncrypt(mode: String, passphrase: String, salt: Option[String])
case class CmdOptions(verbose: Boolean, debug: Boolean)

object CLI extends ParseFileset with ParseEncrypt with ParseOptions {
  val parser = Command[Args](name = "store", header = "Store files")(opts = {
    (fileset, encrypt, options)
      .mapN { (fileset, encrypt, options) =>
        Args(fileset, encrypt, options)
      }
  })
}
trait ParseFileset extends UtilsParse {
  def fileset: Opts[CmdFileset] = {
    val chdir = Opts.option[String](short = "C", long = "chdir", help = "Optional salt.").withDefault(".")
    val label = Opts.option[String](short = "l", long = "label", help = "Fileset label. Optional if only one file is specified.").withDefault("")
    val files = Opts.arguments[String]("files")
    (chdir, label, files)
      .mapN { (chdir, label, files) =>
        CmdFileset(chdir, optionOfString(label), files.toList)
      }
  }
}
trait ParseEncrypt extends UtilsParse {
  def encrypt: Opts[CmdEncrypt] = {
    val mode       = Opts.option[String](short = "m", long = "mode",       help = "Mode of operation. Defaults to FS.").withDefault("FS")
    val passphrase = Opts.option[String](short = "p", long = "passphrase", help = "Passphrase.")
    val salt       = Opts.option[String](short = "s", long = "salt",       help = "Optional salt.").withDefault("")
    (mode, passphrase, salt)
      .mapN { (mode, passphrase, salt) =>
        CmdEncrypt(mode, passphrase, optionOfString(salt))
      }
  }
}
trait ParseOptions {
  def options: Opts[CmdOptions] = {
    val verbose = Opts.flag(short = "v", long = "verbose", help = "Verbose.").orFalse
    val debug   = Opts.flag(short = "d", long = "debug",   help = "Debug.").orFalse
    (verbose, debug)
      .mapN { (verbose, debug) => CmdOptions(verbose, debug) }
  }
}
trait UtilsParse {
  def optionOfString(s: String): Option[String] = if (s.trim.length == 0) None else Some(s.trim)
  def listOfString(s: String, regex: String = ","): Seq[String] = s.split(regex)
}
```
